x/ibc/handler.go

```diff
@@ -37,22 +38,162 @@ func NewHandler(k Keeper) sdk.Handler {
 
 		// IBC channel msgs
 		case channel.MsgChannelOpenInit:
-			return channel.HandleMsgChannelOpenInit(ctx, k.ChannelKeeper, msg)
+			// Lookup module by port capability
+			module, portCap, ok := k.PortKeeper.LookupModuleByPort(ctx, msg.PortID)
+			if !ok {
+				return nil, sdkerrors.Wrap(port.ErrInvalidPort, "could not retrieve module from portID")
+			}
+			res, cap, err := channel.HandleMsgChannelOpenInit(ctx, k.ChannelKeeper, portCap, msg)
+			if err != nil {
+				return nil, err
+			}
+			// Retrieve callbacks from router
+			cbs, ok := k.Router.GetRoute(module)
+			if !ok {
+				return nil, sdkerrors.Wrapf(port.ErrInvalidRoute, "route not found to module: %s", module)
+			}
+			err = cbs.OnChanOpenInit(ctx, msg.Channel.Ordering, msg.Channel.ConnectionHops, msg.PortID, msg.ChannelID, cap, msg.Channel.Counterparty, msg.Channel.Version)
+			if err != nil {
+				return nil, err
+			}
+
+			return res, nil
```

Comment:
nit, can we move each of the cases to a private function to avoid bloating the `switch` and for ease of reading?